x-pack/platform/plugins/shared/onechat/server/routes/mcp.ts

```diff
@@ -0,0 +1,205 @@
+        xsrfRequired: false,
+        availability: {
+          stability: 'experimental',
+        },
+      },
+    })
+    .addVersion(
+      {
+        version: '2023-10-31',
+        validate: {
+          request: { body: schema.object({}, { unknowns: 'allow' }) },
+        },
+      },
+      wrapHandler(async (_, request, response) => {
+        let transport: KibanaMcpHttpTransport | undefined;
+        let server: McpServer | undefined;
+
+        try {
+          transport = new KibanaMcpHttpTransport({ sessionIdGenerator: undefined, logger });
+
+          // Instantiate new MCP server upon every request, no session persistence
+          server = new McpServer({
+            name: MCP_SERVER_NAME,
+            version: MCP_SERVER_VERSION,
+          });
+
+          const { tools: toolService } = getInternalServices();
+
+          const registry = toolService.registry.asScopedPublicRegistry({ request });
+          const tools = await registry.list({});
+
+          // Expose tools scoped to the request
+          for (const tool of tools) {
+            server.tool(
+              tool.id,
+              tool.description,
+              tool.schema.shape,
+              async (args: { [x: string]: any }) => {
+                const toolResult = await tool.execute({ toolParams: args });
+                return {
+                  content: [{ type: 'text' as const, text: JSON.stringify(toolResult) }],
+                };
+              }
+            );
+          }
+
+          request.events.aborted$.subscribe(async () => {
+            await transport?.close().catch((error) => {
+              logger.error('MCP Server: Error closing transport', { error });
+            });
+            await server?.close();
+          });
+
+          await server.connect(transport);
+
+          return await transport.handleRequest(request, response);
+        } catch (error) {
+          logger.error('MCP Server: Error handling request', { error });
+          try {
+            await transport?.close();
+          } catch (closeError) {
+            logger.error('MCP Server: Error closing transport during error handling', {
+              error: closeError,
+            });
+          }
+          if (server) {
+            try {
+              await server.close();
+            } catch (closeError) {
+              logger.error('MCP Server: Error closing server during error handling', {
+                error: closeError,
+              });
+            }
+          }
+
+          logger.error('MCP Server: Error handling request', { error });
+          return response.customError({
+            statusCode: 500,
+            body: {
+              message: `Internal server error: ${error}`,
+              attributes: {
+                code: INTERNAL_ERROR,
+              },
+            },
+          });
+        }
+      })
+    );
+
+  router.versioned
+    .get({
+      path: MCP_SERVER_PATH,
+      security: {
+        authz: {
```

Comment:
Any reasons to define those `GET` and `DELETE` handlers? is that part of the MCP spec to return 405 here?

Reply:
Yes, it's part of the spec. E.g. if GET returns 2xx it means that SSE is supported. Not sure about DELETE but all ([python](https://github.com/modelcontextprotocol/python-sdk/blob/main/src/mcp/server/streamable_http.py#L273), tsx) implementations of the protocol implement it like that.